src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs

```diff
@@ -46,96 +42,174 @@ public class Settings : ISettings
         {
-            if (string.IsNullOrEmpty(root))
+            if (string.IsNullOrEmpty(sectionName))
             {
-                throw new ArgumentException("root cannot be null or empty", nameof(root));
+                throw new ArgumentNullException(nameof(sectionName));
             }
 
-            if (string.IsNullOrEmpty(fileName))
+            if (item == null)
             {
-                throw new ArgumentException(Resources.Argument_Cannot_Be_Null_Or_Empty, nameof(fileName));
+                throw new ArgumentNullException(nameof(item));
             }
 
-            if (!FileSystemUtility.IsPathAFile(fileName))
+            // Check if set is an update
+            if (_computedSections.TryGetValue(sectionName, out var section))
             {
-                throw new ArgumentException(Resources.Settings_FileName_Cannot_Be_A_Path, nameof(fileName));
+                if (section.TryUpdateChildItem(item, isBatchOperation))
+                {
+                    return true;
+                }
             }
 
-            Root = root;
-            FileName = fileName;
-            XDocument config = null;
-            ExecuteSynchronized(() => config = XmlUtility.GetOrCreateDocument(CreateDefaultConfig(), ConfigFilePath));
-            ConfigXDocument = config;
-            IsMachineWideSettings = isMachineWideSettings;
-            CheckConfigRoot();
+            var outputSettingsFile = GetOutputSettingsFileForSection(sectionName);
+
+            // Set is an add
+            return outputSettingsFile != null && outputSettingsFile.RootElement.AddItemInSection(sectionName, item.Copy(), isBatchOperation);
         }
 
         public event EventHandler SettingsChanged = delegate { };
 
-        public IEnumerable<ISettings> Priority
+        public Settings(string root)
+            : this(new SettingsFile(root)) { }
+
+        public Settings(string root, string fileName)
+            : this(new SettingsFile(root, fileName)) { }
+
+        public Settings(string root, string fileName, bool isMachineWide)
+            : this(new SettingsFile(root, fileName, isMachineWide)) { }
+
+        private Settings(SettingsFile settingsHead)
+        {
+            _settingsHead = settingsHead;
+
+            var curr = _settingsHead;
+            while (curr != null)
+            {
+                curr.SettingsChanged += (_, __) => { RefreshSettingValues(); };
+                curr = curr.Next;
+            }
+
+            MergeSettingFilesValues();
+        }
+
+        private void RefreshSettingValues()
+        {
+            MergeSettingFilesValues();
+            SettingsChanged?.Invoke(this, EventArgs.Empty);
+        }
+
+
+        private void MergeSettingFilesValues()
+        {
+            var computedSections = new Dictionary<string, SettingsSection>();
+
+            var curr = _settingsHead;
+            while (curr != null)
```

Comment:
As I mentioned above... this adds unnecessary additional perf cost of merge all the settings files over and over again each time you change a single item in any of the file.

Reply:
I don't believe this perf cost is unnecessary. This is only done on each set operation (i.e. when a settings file is saved). By doing this we gain the freedom in the API to interact with either computed settings or directly with a settings file without having a wrapper in between where you are limited to the exposed interface of that given wrapper. 

The previous code computed and merged all settings files *on every get operation*, this update changes that to compute everything on sets because set operations are way less common than get operations.